### PR TITLE
Add lax.tridiagonal_solve(..), lowering to cusparse<T>gtsv2() on GPU.

### LINKED ISSUE #6830: Tridiagonal solver for CUDA
It would be wonderful if JAX had a tridiagonal solver, perhaps modeled after `scipy.linalg.solve_banded`. There is already a CUDA kernel for this, it would just need to be wrapped for GPU: https://docs.nvidia.com/cuda/cusparse/index.html#gtsv2. I leave it up to you to figure out how best to do this on TPU.

### PR DESCRIPTION
**Pull request by tomhennigan**
Fixes #6830.

### CODE REVIEW

**Comment by shoyer** [comment]
This looks awesome!

Two minor suggestions:
1. Maybe put this in `_src/lax/linalg.py` along with the other linear algebra routines?
2. If you want to define auto-diff rules, I think `lax.custom_linear_solve` would be an easy way to do that. (Of course, this is also definitely fine to leave for later!)

File: jax/_src/lax/lax.py [hunk 1]
```diff
@@ -6913,3 +6914,45 @@ def _canonicalize_axis(axis, num_dims):
   if axis < 0:
     axis = axis + num_dims
   return axis
+
```
Comment (hawkinsp) [inline jax/_src/lax/lax.py:6917]: I'd put this primitive in `jax/_src/lax/linalg.py`.

File: jax/_src/lax/lax.py [hunk 2]
```diff
@@ -6913,3 +6914,45 @@ def _canonicalize_axis(axis, num_dims):
...
+      Note that ``dl[0] = 0``.
+    d: The middle diagnoal of A: ``d[i]  := A[i, i]`` for i in ``[0,m)``.
+    du: The upper diagonal of A: ``du[i] := A[i, i+1]`` for i in ``[0,m)``.
+      Note that ``dl[m - 1] = 0``.
+    B: Right hand side matrix.
+
+  Returns:
+    Solution ``X`` of tridiagonal system.
+  """
+  assert dl.ndim == d.ndim == du.ndim == 1, "dl, d and du must be vectors"
```
Comment (hawkinsp) [inline jax/_src/lax/lax.py:6947]: It's somewhat impolite to use `assert` for user-facing errors.

File: jaxlib/cusparse.py [hunk 3]
```diff
@@ -287,3 +287,16 @@ def coo_matmat(c, data, row, col, B, *, shape, transpose=False, compute_dtype=No
...
   )
   return _ops.GetTupleElement(out, 0)
+
+
+def gtsv2(c, dl, d, du, B, *, m, n, ldb, t):
+  """Calls `cusparse<t>gtsv2(dl, d, du, B, m, n, ldb)`."""
+  dl_shape, d_shape, du_shape, B_shape = map(c.get_shape, (dl, d, du, B))
+  return xla_client.ops.CustomCallWithLayout(
+      c,
+      b"cusparse_gtsv2_" + (b"f32" if (t == np.float32) else b"f64"),
```
Comment (hawkinsp) [inline jaxlib/cusparse.py:297]: Would you mind also supporting the complex and complex double cases as well? Also fine to leave a TODO.

**Review by hawkinsp — CHANGES_REQUESTED** [review]
Looks pretty good.

File: jax/_src/lax/lax.py (hunk 1 above)
Comment (tomhennigan) [reply jax/_src/lax/lax.py]: Done.

File: jax/_src/lax/lax.py (hunk 2 above)
Comment (tomhennigan) [reply jax/_src/lax/lax.py]: Upgraded to ValueError.

File: jaxlib/cusparse.py (hunk 3 above)
Comment (tomhennigan) [reply jaxlib/cusparse.py:297]: Have left a TODO since i'm struggling to find time to write tests for this.

**Comment by tomhennigan** [comment]
Thanks both!

> 1. Maybe put this in `_src/lax/linalg.py` along with the other linear algebra routines?

Done.

> 2. If you want to define auto-diff rules, I think `lax.custom_linear_solve` would be an easy way to do that. (Of course, this is also definitely fine to leave for later!)

Have left a TODO 😄

File: jax/_src/lax/linalg.py [hunk 4]
```diff
@@ -1350,3 +1353,61 @@ def svd_batching_rule(batched_args, batch_dims, full_matrices, compute_uv):
...
+      Note that ``dl[0] = 0``.
+    d: The middle diagnoal of A: ``d[i]  := A[i, i]`` for i in ``[0,m)``.
+    du: The upper diagonal of A: ``du[i] := A[i, i+1]`` for i in ``[0,m)``.
+      Note that ``dl[m - 1] = 0``.
+    b: Right hand side matrix.
+
+  Returns:
+    Solution ``X`` of tridiagonal system.
+  """
+  if dl.ndim != 1 or d.ndim != 1 or du.ndim != 1:
```
Comment (hawkinsp) [inline jax/_src/lax/linalg.py:1387]: Nit: these checks should probably go in the abstract_eval rule.

**Review by hawkinsp — APPROVED** [review]

**Comment by joglekara** [comment]
This is awesome indeed! 

Any chance you might take on the stridedBatch version that solves `n_batch` linear systems? https://docs.nvidia.com/cuda/cusparse/index.html#gtsv2stridedbatch_bufferSize

**Pull request merged**